File: C/leetCode25.c
#include<stdio.h>
#include<malloc.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* reverseKGroup(struct ListNode* head, int k){
	
	if (k==1) return head;
	
	struct ListNode *virtualHead,*h,*t,*p,*q,*last;
	
	virtualHead = (struct ListNode *)malloc(sizeof(struct ListNode));
	virtualHead->val = -1;
	virtualHead->next = head;
	h = head;
	t = head;
	last = virtualHead;
	
	int i;
	for (i=1;i<k && t; i++)
		t = t->next;
	
	while (t)
	{
		p = h;
		h = h->next;
		q = h->next;
		p->next = t->next;
		
		while (h!=t)
		{
			h->next = p;
			p = h;
			h = q;
			q = q->next;
		}
		
		t = last->next;
		last->next = h;
		h->next = p;
		last = t;
		t = q;
		h = q;
		
		for (i=1;i<k && t; i++)
			t = t->next;
	}
		
	head = virtualHead->next;
	return head;
	
}


int main(void)
{
	int num[10] = {1,2,3,4};
	int numSize = 4;
	int i;
	
	struct ListNode * head, *p;
	head = (struct ListNode*) malloc(sizeof(struct ListNode));
	head->next = NULL;
	p = head;
	for (i=0;i<numSize;i++)
	{
		p->next = (struct ListNode*) malloc(sizeof(struct ListNode));
		p = p->next;
		
		p->next = NULL;
		p->val = num[i];
	}
	head = head->next;
	
	head = reverseKGroup(head, 1);
	while (head)
	{
		printf("%d ", head->val);
		head = head->next;
	}
}

File: LeetCode25.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head, k):
        if k == 1:
            return head
        virtualHead = ListNode()
        virtualHead.next = head
        last = virtualHead
        start = head
        p = start
        cnt = 0
        while p != None:
            p = p.next
            cnt += 1
            if cnt == k:
                end = p 

                t = start.next
                l = start
                start.next = end

                while (t!=end):
                    t_next = t.next
                    t.next = l
                    l = t
                    t = t_next
                
                last.next = l
                last = start
                start = end 

                cnt = 0 
                    

        return virtualHead.next

def printList(head):
    p = head
    while (p!=None):
        print(p.val)
        p = p.next

def buildList(List):
    last = ListNode
    q = last
    for i in List:
        p = ListNode(i)
        q.next = p 
        q = p
    return last.next 


if __name__ == "__main__":
    arr = [1,2,3,4,5]
    k = 6
    head = buildList(arr)

    solution = Solution()
    ansHead = solution.reverseKGroup(head, k)
    printList(ansHead)
